# 🧠 Baseline Model: Implied Probability as Predictor

In this notebook, we build a simple baseline model for horse race outcomes using **implied probability** derived from market odds.

## 🎯 Goal

To establish a benchmark model that predicts the probability of a horse winning — using **only its implied win probability** from SP odds.

This baseline:

- Provides a strong starting point for future models
- Helps quantify how much “edge” more complex models might offer
- Reflects the idea that **market odds already encode substantial information**

## ⚙️ Approach

- Use the `implied_prob` column (already included in the processed dataset)
- Treat this as the model's predicted probability of winning
- Evaluate performance using basic metrics (e.g. log loss, Brier score, calibration)

This simple model reflects the **market consensus** — and sets the bar we must beat to justify more complex predictors.


###  Step 1: Load Data


In [2]:
import pandas as pd

# Load cleaned data
YEAR = 2019
df = pd.read_csv(f"../data/processed/horses_{YEAR}_clean.csv")

# Filter to only horses that finished
df = df[df['did_finish'] == 1].copy()

# Confirm shape
print(f"✅ Loaded {len(df):,} race entries (finished only)")


✅ Loaded 163,816 race entries (finished only)


### 🎯 Step 2: Define a Baseline Model Using Implied Probability

Before building complex models, we start with a **baseline** — a simple, reasonable model that future models must outperform.

In this case, our baseline model will use the **implied probability** (derived from betting odds) as its predicted win chance.

This allows us to:
- Test how well the market predicts race outcomes
- Compare future models to the efficiency of bookmaker odds
- Establish a lower benchmark for model performance

We'll define:
- `y_true`: Whether the horse **actually won** (1) or not (0)
- `y_pred`: The model's predicted probability — here, just the `implied_prob`

This baseline model is interpretable, grounded in real-world betting behaviour, and surprisingly strong in practice.


In [4]:
# Define ground truth and model prediction
y_true = df['position'] == 1         # 1 if horse won, 0 otherwise
y_pred = df['implied_prob']          # Our baseline model's prediction

In [7]:
from sklearn.metrics import log_loss, brier_score_loss

df['is_winner'] = (df['position'] == 1).astype(int)

# Define target and prediction
y_true = df['is_winner']
y_pred = df['implied_prob']

# Calculate metrics
logloss = log_loss(y_true, y_pred)
brier = brier_score_loss(y_true, y_pred)

print(f"📉 Log Loss: {logloss:.4f}")
print(f"🎯 Brier Score: {brier:.4f}")


📉 Log Loss: 0.2812
🎯 Brier Score: 0.0817


### 📊 Baseline Model Evaluation – Implied Probabilities vs. Actual Outcomes

To assess how well market-implied probabilities predict actual race results, we evaluated them using two standard probabilistic scoring metrics:

---

#### 📉 Log Loss: **0.2812**
- A measure of how confidently and correctly probabilities predict binary outcomes (win or not).
- Ranges from 0 (perfect predictions) upwards — lower is better.
- Our value of ~0.28 suggests **relatively well-calibrated** predictions for a baseline model.

#### 🎯 Brier Score: **0.0817**
- Measures the **mean squared difference** between predicted probabilities and actual outcomes (0 or 1).
- Ranges from 0 (perfect calibration) to 1 (worst).
- A score of ~0.08 indicates **low prediction error** across the dataset.

---

### ✅ Takeaways:
- The market is quite effective at predicting outcomes using implied probability — as measured by both log loss and Brier score.
- This serves as a **strong benchmark**: any custom model must perform better than this to be worthwhile.
- It confirms why market odds are commonly used as a **feature or baseline** in predictive racing models.
